Mathlib/Topology/Algebra/Module/Basic.lean

```diff
@@ -2272,34 +2272,153 @@ end ContinuousLinearEquiv
 
 namespace ContinuousLinearMap
 
-variable {R : Type*} {M : Type*} {M₂ : Type*} [TopologicalSpace M] [TopologicalSpace M₂]
+variable {R : Type*} {M M₂ M₃ : Type*}
+  [TopologicalSpace M] [TopologicalSpace M₂] [TopologicalSpace M₃]
 
 section
 
 variable [Semiring R]
-variable [AddCommMonoid M₂] [Module R M₂]
-variable [AddCommMonoid M] [Module R M]
+  [AddCommMonoid M] [Module R M]
+  [AddCommMonoid M₂] [Module R M₂]
+  [AddCommMonoid M₃] [Module R M₃]
+
+/-- A continuous linear map is invertible if it is the forward direction of a continuous linear
+equivalence. -/
+def IsInvertible (f : M →L[R] M₂) : Prop :=
+  ∃ (A : M ≃L[R] M₂), A = f
 
 open Classical in
 /-- Introduce a function `inverse` from `M →L[R] M₂` to `M₂ →L[R] M`, which sends `f` to `f.symm` if
 `f` is a continuous linear equivalence and to `0` otherwise.  This definition is somewhat ad hoc,
 but one needs a fully (rather than partially) defined inverse function for some purposes, including
 for calculus. -/
 noncomputable def inverse : (M →L[R] M₂) → M₂ →L[R] M := fun f =>
-  if h : ∃ e : M ≃L[R] M₂, (e : M →L[R] M₂) = f then ((Classical.choose h).symm : M₂ →L[R] M) else 0
+  if h : f.IsInvertible then ((Classical.choose h).symm : M₂ →L[R] M) else 0
+
+@[simp] lemma isInvertible_equiv {f : M ≃L[R] M₂} : IsInvertible (f : M →L[R] M₂) := ⟨f, rfl⟩
 
 /-- By definition, if `f` is invertible then `inverse f = f.symm`. -/
 @[simp]
 theorem inverse_equiv (e : M ≃L[R] M₂) : inverse (e : M →L[R] M₂) = e.symm := by
-  have h : ∃ e' : M ≃L[R] M₂, (e' : M →L[R] M₂) = ↑e := ⟨e, rfl⟩
-  simp only [inverse, dif_pos h]
-  congr
-  exact mod_cast Classical.choose_spec h
+  simp [inverse]
 
 /-- By definition, if `f` is not invertible then `inverse f = 0`. -/
-@[simp]
-theorem inverse_non_equiv (f : M →L[R] M₂) (h : ¬∃ e' : M ≃L[R] M₂, ↑e' = f) : inverse f = 0 :=
-  dif_neg h
+@[simp] lemma inverse_of_not_isInvertible
+    {f : M →L[R] M₂} (hf : ¬ f.IsInvertible) : f.inverse = 0 :=
+  dif_neg hf
+
+@[deprecated (since := "2024-10-29")] alias inverse_non_equiv := inverse_of_not_isInvertible
+
+@[simp] theorem inverse_zero : inverse (0 : M →L[R] M₂) = 0 := by
+  by_cases h : IsInvertible (0 : M →L[R] M₂)
+  · rcases h with ⟨e', he'⟩
+    simp only [← he', inverse_equiv]
+    ext v
+    apply e'.injective
+    rw [← ContinuousLinearEquiv.coe_coe, he']
+    rfl
+  · exact inverse_of_not_isInvertible h
+
+lemma IsInvertible.comp {g : M₂ →L[R] M₃} {f : M →L[R] M₂}
+    (hg : g.IsInvertible) (hf : f.IsInvertible) : (g ∘L f).IsInvertible := by
+  rcases hg with ⟨N, rfl⟩
+  rcases hf with ⟨M, rfl⟩
+  exact ⟨M.trans N, rfl⟩
+
+lemma IsInvertible.of_inverse {f : M →L[R] M₂} {g : M₂ →L[R] M}
+    (hf : f ∘L g = id R M₂) (hg : g ∘L f = id R M) :
```

Comment:
I think you should just add a convenience constructor for `ContinuousLinearEquiv` with these hypotheses. I'm surprised we don't have it already. That will simplify the proof of this and the next lemma.